File: AspNet472CSharpMVC5/SameSiteCookieManager.cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

using Microsoft.Owin;
using Microsoft.Owin.Infrastructure;

namespace AspNet472CSharpMVC5
{
    public class SameSiteCookieManager : ICookieManager
    {
        private readonly ICookieManager _innerManager;

        public SameSiteCookieManager() : this(new CookieManager())
        {
        }

        public SameSiteCookieManager(ICookieManager innerManager)
        {
            _innerManager = innerManager;
        }

        public void AppendResponseCookie(IOwinContext context, string key, string value,
                                         CookieOptions options)
        {
            CheckSameSite(context, options);
            _innerManager.AppendResponseCookie(context, key, value, options);
        }

        public void DeleteCookie(IOwinContext context, string key, CookieOptions options)
        {
            CheckSameSite(context, options);
            _innerManager.DeleteCookie(context, key, options);
        }

        public string GetRequestCookie(IOwinContext context, string key)
        {
            return _innerManager.GetRequestCookie(context, key);
        }

        private void CheckSameSite(IOwinContext context, CookieOptions options)
        {
            if (options.SameSite == Microsoft.Owin.SameSiteMode.None &&
                                    SameSite.BrowserDetection.DisallowsSameSiteNone(context.Request.Headers["User-Agent"]))
            {
                options.SameSite = null;
            }
        }
    }
}

File: AspNet472VisualBasicMVC5/SameSiteCookieManager.vb
' Copyright (c) .NET Foundation. All rights reserved.
' Licensed under the Apache License, Version 2.0. See License.txt In the project root for license information.
Imports Microsoft.Owin
Imports Microsoft.Owin.Infrastructure

Public Class SameSiteCookieManager
    Implements ICookieManager

    Private ReadOnly _innerManager As ICookieManager

    Public Sub New()
        Me.New(New CookieManager())
    End Sub

    Public Sub New(ByVal innerManager As ICookieManager)
        _innerManager = innerManager
    End Sub

    Public Sub AppendResponseCookie(ByVal context As IOwinContext, ByVal key As String, ByVal value As String, ByVal options As CookieOptions) Implements ICookieManager.AppendResponseCookie
        CheckSameSite(context, options)
        _innerManager.AppendResponseCookie(context, key, value, options)
    End Sub

    Public Sub DeleteCookie(ByVal context As IOwinContext, ByVal key As String, ByVal options As CookieOptions) Implements ICookieManager.DeleteCookie
        CheckSameSite(context, options)
        _innerManager.DeleteCookie(context, key, options)
    End Sub

    Public Function GetRequestCookie(ByVal context As IOwinContext, ByVal key As String) As String Implements ICookieManager.GetRequestCookie
        Return _innerManager.GetRequestCookie(context, key)
    End Function

    Private Sub CheckSameSite(ByVal context As IOwinContext, ByVal options As CookieOptions)
        If options.SameSite = Microsoft.Owin.SameSiteMode.None AndAlso SameSite.BrowserDetection.DisallowsSameSiteNone(context.Request.Headers("User-Agent")) Then
            options.SameSite = Nothing
        End If
    End Sub
End Class
